### PR TITLE
ET-354-add-vitals

### PR DESCRIPTION
**Pull request by sharlotta93**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Implemented the `addVitals` action to add patient vitals, including validation and API interaction.
- Added unit tests for the `addVitals` action, covering success and error scenarios.
- Defined new constants, fields, and data points related to vitals.
- Enhanced the API client with a new method to handle vitals data.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>client.ts</strong><dd><code>Add mock implementation for addVitals function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/__mocks__/client.ts

<li>Added <code>vitalsResponseExample</code> to the imports.<br> <li> Added <code>addVitals</code> mock function to return <code>vitalsResponseExample</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-f233b91df2b9bf19f5f13ccc7522c18dd4a8085ebdb2812dc8dfab8a04372049">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>constants.ts</strong><dd><code>Add vitals example and response constants</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/__mocks__/constants.ts

- Added `vitalsExample` and `vitalsResponseExample` constants.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-1620308b774afef7cdd55c47a2fc1c5d728f51202540645730dc217894556fb6">+71/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>addVitals.ts</strong><dd><code>Implement addVitals action for patient vitals</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addVitals/addVitals.ts

<li>Implemented <code>addVitals</code> action for adding patient vitals.<br> <li> Included validation and API client interaction.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-a76ca54a7372487d46958ad55f93b354681c10f552a16e75555e40d48751e83f">+89/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>dataPoints.ts</strong><dd><code>Define dataPoints for addVitals action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addVitals/config/dataPoints.ts

- Defined `dataPoints` for vitalsId.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-92ccedf96068c527738c40e477388d05331107f6bdf37fbe0c815ca01102895b">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fields.ts</strong><dd><code>Define fields and validation for addVitals</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addVitals/config/fields.ts

- Defined fields and validation schema for vitals data.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-014b3ff309bfc661b58417e7372e1d2f3efc46ef4dd12241516d50297b58e429">+246/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export fields and dataPoints for addVitals</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addVitals/config/index.ts

- Exported fields and dataPoints from config.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-dd912050ce11489f6ef69cdaf2fc65c39007462636f366e2293308fcba12d49c">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export addVitals action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addVitals/index.ts

- Exported `addVitals` action.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-20e1ece44d4aff265dccf7571bd1d5ddf3376250721fe987924db33f8470f89d">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>client.ts</strong><dd><code>Add addVitals method to API client</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/client.ts

<li>Added <code>addVitals</code> method to <code>ElationDataWrapper</code> and <code>ElationAPIClient</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-65c140fbe1debe73b11455b75ab0aa6aa573b3c47babb49245c0d35c8e62a662">+19/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>vitals.ts</strong><dd><code>Define types and schemas for vitals</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/types/vitals.ts

- Defined types and schemas for vitals input and response.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-5da9f8b308d1c897d478977d4850a27c72ac4eaf9f54cef68defa2999e48b961">+92/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>addVitals.test.ts</strong><dd><code>Add unit tests for addVitals action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addVitals/addVitals.test.ts

<li>Added tests for <code>addVitals</code> action.<br> <li> Included scenarios for success and various error conditions.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/511/files#diff-e17e4c4b0f32bef83535b8998ce6fdf0c8815e45b5b4a01668952d7510a50868">+100/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: extensions/elation/actions/addVitals/config/fields.ts [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    label: 'Practice ID',
+    description:
+      'Practice ID where the vitas are beind measured for the patient',
+    type: FieldType.NUMERIC,
+    required: true,
+  },
+  visitNoteId: {
+    id: 'visitNoteId',
+    label: 'Visit Note ID',
+    description: 'Will only allow unsigned visit notes',
```
Comment (nckhell) [inline extensions/elation/actions/addVitals/config/fields.ts:27]:
```suggestion
    description: 'When provided, the vitals will be attached to the visit note. Note that vitals can only be attached to unsigned visit notes.',
```

File: extensions/elation/actions/addVitals/config/fields.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    type: FieldType.NUMERIC,
+    required: false,
+  },
+  nonVisitNoteId: {
+    id: 'nonVisitNoteId',
+    label: 'Non-Visit Note ID',
+    description: 'Will only allow unsigned non-visit notes',
```
Comment (nckhell) [inline extensions/elation/actions/addVitals/config/fields.ts:34]:
```suggestion
    description: 'When provided, the vitals will be attached to the non-visit note. Note that vitals can only be attached to unsigned non-visit notes.',
```

File: extensions/elation/actions/addVitals/config/fields.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    type: FieldType.STRING,
+    required: false,
+  },
+  weight: {
+    id: 'weight',
+    label: 'Weight',
+    description: 'Patient weight in lbs',
+    type: FieldType.STRING,
+    required: false,
+  },
```
Comment (nckhell) [inline extensions/elation/actions/addVitals/config/fields.ts:66]: The field type `string` will cause issues as data points are strongly typed in Awell. E.g. usually "weight" is collected in a form via a numeric question that results in a numeric data point value type. The builder will not be able to map this data point to this action field as the field type is string...

I'd suggest that for each measurement in Elation, we check what the logical data type would be in our system.

See https://www.loom.com/share/b0dc827e81ff4cff9488fbbbe9b83e52

File: extensions/elation/actions/addVitals/addVitals.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  fields as elationFields,
+  dataPoints,
+} from './config'
+
+// Helper function to create measurement objects with optional notes
+const createMeasurement = (
+  value?: string,
+  note?: string
+): Array<z.infer<typeof measurementInputSchema>> | undefined => {
+  if (isNil(value)) {
```
Comment (nckhell) [inline extensions/elation/actions/addVitals/addVitals.ts:21]: From experience, an `isEmpty` check on the value is safer.

**Why?**
Once you configured an action field with a static value and you remove it, the value is `""` and not `undefined`.

**Review by nckhell — CHANGES_REQUESTED** [review]

File: extensions/elation/actions/addVitals/config/fields.ts (hunk 3 above)
Comment (sharlotta93) [reply extensions/elation/actions/addVitals/config/fields.ts:66]: omg, I can't believe I didn't think of that. Very good point and thank you for flagging this. A valuable lesson for sure.

**Pull request merged**